src/mono/mono/utils/memfuncs.c

```diff
@@ -292,4 +292,75 @@ mono_determine_physical_ram_size (void)
 #else
 	return 134217728;
 #endif
-}
\ No newline at end of file
+}
+
+guint64
+mono_determine_physical_ram_available_size (void)
+{
+#if defined (TARGET_WIN32)
+	MEMORYSTATUSEX memstat;
+
+	memstat.dwLength = sizeof (memstat);
+	GlobalMemoryStatusEx (&memstat);
+	return (guint64)memstat.ullAvailPhys;
+
+#elif defined (__NetBSD__)
+	struct vmtotal vm_total;
+	guint64 page_size;
+	int mib[2];
+	size_t len;
+
+	mib[0] = CTL_VM;
+	mib[1] = VM_METER;
+
+	len = sizeof (vm_total);
+	sysctl (mib, 2, &vm_total, &len, NULL, 0);
+
+	mib[0] = CTL_HW;
+	mib[1] = HW_PAGESIZE;
+
+	len = sizeof (page_size);
+	sysctl (mib, 2, &page_size, &len, NULL, 0);
+
+	return ((guint64) vm_total.t_free * page_size) / 1024;
+#elif defined (__APPLE__)
+	mach_msg_type_number_t count = HOST_VM_INFO_COUNT;
+	mach_port_t host = mach_host_self ();
+	vm_size_t page_size;
+	vm_statistics_data_t vmstat;
+	kern_return_t ret;
+	do {
+		ret = host_statistics (host, HOST_VM_INFO, (host_info_t)&vmstat, &count);
+	} while (ret == KERN_ABORTED);
+
+	if (ret != KERN_SUCCESS) {
+		g_warning ("Mono was unable to retrieve memory usage!");
+		return 0;
+	}
+
+	host_page_size (host, &page_size);
+	return (guint64) vmstat.free_count * page_size;
+
+#elif defined (HAVE_SYSCONF)
```

Comment:
Can't we try `/proc/meminfo` first?

Reply:
Judging by https://github.com/dotnet/runtime/issues/13371 the answer is yes, we should.

Reply:
As this PR fixes a number of issues and is ready to go, I am going to merge it, and create another issue to fix that.

Reply:
Since this fixes a number of issues and is ready to, I am incline to merge it and create a seperate issue for this. It is going to take me a little bit to add parsing for /proc and make sure it is consistent on different linuxes.

Reply:
Sure, though i'd like for it to not fall off the radar since the exact issue has already been hashed out in CoreCLR. It shouldn't be too complicated, since if MemAvailable isn't present in `/proc/meminfo` you can just fall back to sysconf. CoreCLR's implementation, for reference: https://github.com/dotnet/runtime/blob/490551c221be0683039cebc1e5da5b12b3527c12/src/coreclr/src/gc/unix/gcenv.unix.cpp#L823-L854